### ISSUE TITLE
{optimize  = true} doesn't work with closures (even without updating variables)?

### ISSUE THREAD

**Issue by oal17** [twitter/torch-autograd#129]
I see that there's some new documentation re optimized derivatives in the readme, but I don't see how it applies to examples like this one:

```lua 
local d = require "autograd"

function test()
    local net = nn.Sequential()
        :add(nn.Linear(2, 1))
        :add(nn.Sigmoid())
    
    local netFn, netParams = d.functionalize(net) 

    local f = function(x)
        return torch.sum(netFn(netParams, x))
    end

    --WORKS WITH NO OPTIMIZATION
    local x = torch.rand(1, 2) 
    local df = d.grad(f)
    local dx = df(x)

    --WORKS WITH OPTIMIZATION AND NO CLOSURE
    local g = function(x, ps)
        return torch.sum(netFn(ps, x))
    end
    local dg = d.grad(g)
    local dx = dg(x, netParams)

    --FAILS WITH OPITMIZATION AND CLOSURE 
    local dfOptimized = d.grad(f, {optimize = true})
    local dxOptimized = dfOptimized(x)
end
```

With the closure, the optimized version gives this error and generated code:
```
return function(locals, rlocals, vlocals, objects)
local nn = require('autograd').nn
local util = require('autograd.util')
local util_fillInPlace = util.fillInPlace
local torch_sum = torch.sum
local model = model
return function(p1)
    locals[1] = objects[1].forward({ 0.4047  0.6308
[torch.DoubleTensor of size 1x2]
, 0.01 *
-9.2744
[torch.DoubleTensor of size 1]
}, p1)
    locals[2] = objects[1].backward((util_fillInPlace(rlocals[1], locals[1], 1)), { 0.4047  0.6308
[torch.DoubleTensor of size 1x2]
, 0.01 *
-9.2744
[torch.DoubleTensor of size 1]
}, p1)
    locals[3] = torch_sum(locals[1])
    return locals[2][2], locals[3]
end
end

[string "return function(locals, rlocals, vlocals, obj..."]:8: '}' expected near '0.6308'
```

Not sure if this is an inherent issue with the optimized version of  `grad`, or if there is a way around it, but thought I would check. Thanks!

**Comment by alexbw** [comment]
With optimized mode, closures are "frozen" after the first time you run the function. Behind the scenes, we cache the function so that we don't have to recompile it again. This has the unfortunate side-effect of ignoring all changes to the closed-over variable in the future, which in this case includes the parameters of your neural net.

In general, in optimized mode
 * Do NOT use closures, they aren't safe.
 * Do NOT use variables for control flow (if-statements, for-loops) that you did not DIRECTLY pass in.

**Comment by oal17** [comment]
Thanks for your quick response! I understand the freezing issue, but I wasn’t sure why it would cause problems cases like this. Given that I’m taking the gradient wrt the network’s input, rather than it’s parameters, it seems like it should be ok for the parameters to be frozen at the initial values they get assigned?

One other note: no idea if there’s any possibility for second-order gradients to be supported in non-optimized mode, but this is definitely a feature I would find useful.

**Comment by alexbw** [comment]
Don't you want to update the parameters?

Yeah, the cool features of autograd are unfortunately split non-uniformly
between direct and optimized mode. I'm working to fix this, but I do not
have an ETA.

**Comment by oal17** [comment]
That sounds right - the numbers in the generated code are indeed the same as the parameter tensor.

**Comment by alexbw** [comment]
Ok, cool. So, pass the parameters in as an argument, and you should be good to go.
